### ISSUE TITLE
Default props behavior changed between v3 and v4

### ISSUE THREAD

**Issue by TimonVS** [styled-components/styled-components#2246]
The behavior of how `defaultProps` are handled has changed between s-c v3 and v4, I'm not exactly sure if this change was intentional but it was not apparent in the changelog, so that's why I'm filing it :).

Anyways, in v3 I could have a `Text` component with some default props such as the font. I could then create a `Heading` component which is based on `Text` but may provide some different default props. In v3, the following example works fine and both the `Heading` and `Text` components always have a default `font` prop. In v4 however, the `defaultProps` of the `Heading` overwrites all the default props of `Text`. If I don't define any `defaultProps` for `Heading` it seems to work fine.

```
const Text = styled.span`
  font-family: ${({ font }) => {
    if (!font) console.error('FONT PROP NOT FOUND!')
    return font
  }};
`

Text.defaultProps = {
  font: 'sans-serif'
}

const Heading = styled(Text)`
  display: block;
  font-size: 38px;
`

Heading.defaultProps = {}
```

## Environment

`npx envinfo --system --binaries --npmPackages styled-components,babel-plugin-styled-components --markdown --clipboard`

N/A, see reproduction.

## Reproduction

https://www.webpackbin.com/bins/-KeeZCr0xKfutOfOujxN
https://codesandbox.io/s/rkmNRByE4

The exact same demo in two different versions of Styled Components:

s-c v3, `Heading` and `Text` are both using `sans-serif`: https://codesandbox.io/s/6wmwk9l8kk
s-c v4, the `font-family` for `Heading` is undefined because the `font` prop is undefined: https://codesandbox.io/s/9jlw0zq5yo

## Expected Behavior

The `Heading` component should inherit the `defaultProps` from the `Text` component like it did in v3. I think in v3 the `defaultProps` would be merged?

![image](https://user-images.githubusercontent.com/876666/49029693-83ec4f00-f1a5-11e8-88ef-0d902e58c475.png)

## Actual Behavior

The `defaultProps` of `Heading` overwrites the `defaultProps` of the `Text` component even though `Heading` is based on the `Text` component.

![image](https://user-images.githubusercontent.com/876666/49029715-923a6b00-f1a5-11e8-885b-4b7e83b136d1.png)

**Comment by probablyup** [comment]
So I think what's happening here is we don't attempt to fold the defaultProps together. In v3 each styled component layer was actually rendered separately, whereas in v4 there is only 1 component rendered no matter how many `styled` override layers there might be.

I'm open to changing this, the only edge case would probably be `style` which has special handling rules. In general though, it's probably better to use `attrs` over `defaultProps` as those do get merged appropriately.

**Comment by probablyup** [comment]
cc @styled-components/core

**Comment by oomathias** [comment]
I'm also looking for a way to define attrs or defaults props that are overridable by the end user and/or a styled parent.

defaultProps nor attrs handle this case:

```js
const Box = styled(X).attrs({
  display: 'block'
})``
// Here display is not overridable

const Box = styled(X).attrs({
  display: props => props.display || 'block'
})``
const Flex = styled(Box).attrs({
  display: props => props.display || 'flex'
})``
// Here it is, but Flex is display 'block' or props.display, it's never 'flex'.
```

You can see an example here: https://codesandbox.io/embed/ylz0jwq1v

**Comment by oomathias** [comment]
I made more tests, dynamic `attrs` works as I would expect with `4.1.1` but doesn't with `4.1.2`.

I used the `.attrs(props => ({}))` way this time.

Codesandbox demo (same code):
- [`4.1.1`](https://codesandbox.io/embed/3x5n7qknp5)
- [`4.1.2`](https://codesandbox.io/embed/0pv42yyvnl)

I guess this change was introduced by #2210

_please don't mind my `dynamic` function, it transforms `display: block` into `display: props.display || 'block'`_

**Comment by Kovensky** [comment]
It looks like this loop overrides anything you have in `props` with whatever was in `attrs` while evaluating the styles: https://github.com/styled-components/styled-components/blob/a41dfd0537f96b2a8572409382faedcdc936849e/src/models/StyledComponent.js#L212-L213

`attrs` does correctly behave like `defaultProps` but only for passing props down to the children; when evaluating the styles it's currently overriding props.

**Comment by kristojorg** [comment]
Do we have a path forward here? Default prop folding is important for me as well while using styled-system. Default props should work the same as props, only they're being defined at build time instead of runtime.

**Comment by oomathias** [comment]
I've made a PR to close this issue https://github.com/styled-components/styled-components/pull/2260 and bring back defaultProps folding. I'm still waiting for feedback, but IMO it LGTM.

ps: They are not being defined at built time, even in React.

**Comment by kristojorg** [comment]
Haha I deleted it before I say your reply because I found your PR : )

**Comment by clentfort** [comment]
Hey, the problem this issue broke us in production. :sob:  Unfortunately there is no warning to be found in the release notes or the README to warn future upgrade.

Anyhow is there anyway we can help to get #2260  merged?

**Comment by njsfield** [comment]
@clentfort same for us in production :(
@probablyup & @oomathias #2260 has been stuck in review for several months, would you be able to revisit this please, and let us know if there's anyway we can help to move along?

**Comment by oomathias** [comment]
@njsfield I'm also waiting. It's a sensitive merge and the core team is missing time to review these kinds of PR.
If any of you had the chance to test #2260, giving feedback would help.

**Comment by probablyup** [comment]
Released in v4.3.0
